SQL Expressions: Add JSON support

kylebrandt opened the pull request:
**What is this feature?**

Support json column types.

Before (GH Datasource):

![image](https://github.com/user-attachments/assets/9fb587a0-705e-425f-86d7-e78b81079b9a)

Now:

![image](https://github.com/user-attachments/assets/746a4b83-3b82-4754-a78c-f26df9e61d01)


**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

kylebrandt committed:
SQL Expressions: Support JSON

kylebrandt committed:
SQL Expressions: Support JSON

kylebrandt committed:
allow json funcs, and fix json conversion in next

kylebrandt committed:
allow json funcs, and fix json conversion in next

kylebrandt committed:
Merge branch 'kb/sqlExprJSON' of github.com:grafana/grafana into kb/sqlExprJSON

samjewell commented on pkg/expr/sql/frame_db_conv.go:
```diff
@@ -159,6 +165,34 @@ func fieldValFromRowVal(fieldType data.FieldType, val interface{}) (interface{},
+			return raw, nil
+	
+		default:
+			return nil, fmt.Errorf("unexpected type for JSON field: %T", val)
```
Surprised to see `%T` here for `val`, when we use `%v` for val four lines below this. I wonder if we can use the same in both places? I don't mind which we choose.

samjewell commented on pkg/expr/sql/parser_allow.go:
```diff
@@ -210,6 +210,12 @@ func allowedFunction(f *sqlparser.FuncExpr) (b bool) {
 		return
 
+	// JSON functions
+	case "json_extract", "json_unquote", "json_contains",
+		"json_object", "json_array", "json_set", "json_remove",
+		"json_length", "json_search", "json_type":
```
I'd like to see us TDD adding these, personally. I'd like to see a test case in `parser_allow_test`, and for that test-case to be runnable in SQL-Fiddle, without and dependency on any pre-existing table or column.

samjewell commented on pkg/expr/sql/db_test.go:
```diff
@@ -205,6 +206,80 @@ func TestErrorsFromGoMySQLServerAreFlagged(t *testing.T) {
+	resultFrame, err := sqlDB.QueryFrames(context.Background(), "json_test", query, data.Frames{expectedFrame})
+	require.NoError(t, err)
+
+	// Ensure consistent names for comparison
+	resultFrame.Name = expectedFrame.Name
+	resultFrame.RefID = expectedFrame.RefID
```
Not the hugest fan of this part, I'll be honest. I'd rather simply test for the things that we do expect to match, rather than mutating the result from the database before we do a comparison.

~~We're passing a `data.FrameTestCompareOptions()...` here, right? Could we tell cmp.Diff to ignore these fields with that, maybe 🤔 ?~~

I would personally omit `Name` and `RefID` from expectedFrame, or set them to what we're actually expecting to get back, when we create that object. And maybe the object at the top should be named `InputFrame`

Same comment applies to the next test in this file too.

kylebrandt committed:
lint

samjewell commented on pkg/expr/sql/db_test.go:
```diff
@@ -205,6 +206,80 @@ func TestErrorsFromGoMySQLServerAreFlagged(t *testing.T) {
+		RefID: "json_test",
+		Name:  "json_test",
+		Fields: []*data.Field{
+			data.NewField("id", nil, []*int64{p(int64(1)), p(int64(2))}),
+			data.NewField("payload", nil, []*json.RawMessage{
+				p(json.RawMessage(`{"foo":1}`)),
+				p(json.RawMessage(`{"bar":"baz"}`)),
+			}),
```
None of these values are null. So I wouldn't bother making them nullable personally.

```suggestion
			data.NewField("id", nil, []int64{1, 2}),
			data.NewField("payload", nil, []json.RawMessage{
				json.RawMessage(`{"foo":1}`),
				json.RawMessage(`{"bar":"baz"}`),
			}),
```

samjewell commented on pkg/expr/sql/db_test.go:
```diff
@@ -205,6 +206,80 @@ func TestErrorsFromGoMySQLServerAreFlagged(t *testing.T) {
+		RefID: "A",
+		Name:  "A",
+		Fields: []*data.Field{
+			data.NewField("title", nil, []*string{
+				p("Bug report"),
+				p("Feature request"),
+			}),
+			data.NewField("labels", nil, []*json.RawMessage{
+				p(json.RawMessage(`["type/bug", "priority/high"]`)),
+				p(json.RawMessage(`["type/feature", "priority/low"]`)),
+			}),
```
Again, none of these are null. Would this work?

```suggestion
			data.NewField("title", nil, []string{"Bug report", "Feature request"}),
			data.NewField("labels", nil, []json.RawMessage{
				json.RawMessage(`["type/bug", "priority/high"]`),
				json.RawMessage(`["type/feature", "priority/low"]`),
			}),
```

`expected` will probably need to change to match too

samjewell reviewed (approved):
Thanks for this. Great to see this going in - thanks for the PR description, and also for the tests 👏 

I asked an LLM to review, and got some extra thoughts:

**Security/Performance:**
- Should we put size-limits on JSON documents? To prevent OOMing.

**Testing:**
- LLM suggested we test more than just the `json_contains` function, and personally I agree - I think we should test all the functions we're adding.
- It asked about testing for NULL values and for nested JSON structures. Maybe we don't need this, I'm not sure.

samjewell commented on pkg/expr/sql/db_test.go:
```diff
@@ -205,6 +206,80 @@ func TestErrorsFromGoMySQLServerAreFlagged(t *testing.T) {
+		},
+	}
+
+	sqlDB := &DB{}
```
In previous tests this line would say:
```suggestion
	db := DB{}
```
I vote for consistency here - would like to see this test and the next one match our previous tests, as much as is possible.

kylebrandt commented on pkg/expr/sql/frame_db_conv.go:
```diff
@@ -159,6 +165,34 @@ func fieldValFromRowVal(fieldType data.FieldType, val interface{}) (interface{},
 
+	case data.FieldTypeJSON, data.FieldTypeNullableJSON:
+		switch v := val.(type) {
+		case types.JSONDocument:
```
Nope, might not be needed, haven't checked yet

samjewell committed:
Remove unused conversions

As we populate the data into the GMS db, we control the data formats
which go in. So we can remove these, as we don't push data in in this
format (yet).

samjewell committed:
Simplify tests

- remove pointers when not testing nullable fields
- align variable names with other tests

samjewell committed:
Stop comparing Name and RefID when using cmp.Diff in tests

samjewell committed:
Make "unsupported type" errors more consistent

samjewell committed:
Remove redundant code-branch

this if-check already exists within the `fT.NullableType()` method,
which is always called further down in this file.

samjewell committed:
Warn on the dataframe is data was coerced

I asked the LLM whether it felt a change was needed here. It said:

> The ignored boolean typically indicates whether the conversion was exact or required coercion. Since we're already checking for errors, we should also verify this conversion status, especially in a database context where data integrity is important.
>
> you could:
> - Add proper logging
> - Consider using a different approach for handling the JSON data
> - Return an error if exact conversion is required

I asked to return an error. My gut tells me we have bigger problems, and
should look to fix code in the rest of the codebase, if we find
ourselves coercing the data here.

So this is what the LLM gave me :-)

samjewell reviewed (approved):
I'm personally very happy with this (assuming that I add the unit-test I've asked for before I merge). Glad to see JSON support going in 🎉

kylebrandt commented on pkg/expr/sql/frame_table.go:
```diff
@@ -90,7 +91,21 @@ func (ri *rowIter) Next(_ *mysql.Context) (mysql.Row, error) {
+				return nil, fmt.Errorf("failed to convert json.RawMessage to JSONDocument: %w", err)
+			}
+			if !converted {
+				return nil, fmt.Errorf("unexpected JSON conversion issue at row %d, column %s: value required type coercion", ri.row, ri.ft.Frame.Fields[colIndex].Name)
```
Not sure I understand the error here. Meaning we were unable to convert?

samjewell replied:
No, see the function signature for `types.JSON.Convert`

![image](https://github.com/user-attachments/assets/4f93d934-1081-49a6-9ce4-8e8249d7ff1f)

But I guess the variable name didn't help here. I'll rename to `inRange`

kylebrandt commented on pkg/expr/sql/frame_db_conv.go:
```diff
@@ -159,8 +162,22 @@ func fieldValFromRowVal(fieldType data.FieldType, val interface{}) (interface{},
+	case data.FieldTypeJSON, data.FieldTypeNullableJSON:
+		switch v := val.(type) {
+		case types.JSONDocument:
+			// Use fmt.Sprintf to serialize JSONDocument to []byte
```
rm old comment

samjewell committed:
Rename variable to `inRange`

To better reflect what it represents

samjewell committed:
Remove old comment

samjewell committed:
Add unit test for JSON functions in SQL expressions

samjewell committed:
Merge branch 'main' into kb/sqlExprJSON

samjewell commented:
Sorry for the noise here - combination of rebasing and merging went wrong. I'll close this PR and start a new one.
